src/rust/engine/src/watch.rs

```diff
@@ -93,26 +91,39 @@ impl InvalidationWatcher {
             let paths: HashSet<_> = ev
               .paths
               .into_iter()
-              .map(|path| {
+              .filter_map(|path| {
                 // relativize paths to build root.
-                let mut paths_to_invalidate: Vec<PathBuf> = vec![];
-                let path_relative_to_build_root = {
-                  if path.starts_with(&canonical_build_root) {
-                    path.strip_prefix(&canonical_build_root).unwrap().into()
-                  } else {
-                    path
-                  }
+                let path_relative_to_build_root = if path.starts_with(&canonical_build_root) {
+                  // Unwrapping is fine because we check that the path starts with
+                  // the build root above.
+                  path.strip_prefix(&canonical_build_root).unwrap().into()
+                } else {
+                  path.clone()
                 };
-                paths_to_invalidate.push(path_relative_to_build_root.clone());
+                if ignorer
+                  .is_ignored_or_child_of_ignored_path(&path_relative_to_build_root, path.is_dir())
```

Comment:
`is_dir` will stat the path to determine whether this is a directory... that's probably not the end of the world, but it's not cheap either.

If this ends up needing another CI burn, it might be good to leave a TODO here mentioning that, and discussing the possibility of maybe using https://docs.rs/notify/5.0.0-pre.2/notify/event/enum.EventKind.html to get this info, or maybe of figuring out whether it is safe to lie to the ignore crate and say that everything is a directory (because I don't think ignore patterns can limit themselves to files).

Reply:
I'll look into the ignore-rs code to see if lying would work.

Reply:
I dont think we can lie to Ignore. We could get erroneous results if the match glob is only supposed to match directories, and we are checking against a path pointing to a file. 

https://github.com/BurntSushi/ripgrep/blob/67c0f576b6e1197c3cf80fe5d53cfd8326cf1d56/crates/ignore/src/gitignore.rs#L258-L266

Reply:
We could probably do something naive like check if the last char of the path is a `/` and get away with it.

Reply:
Unfortunately paths to directory events do not come back with a trailing slash from notify so ☝️ wont work.

Reply:
> I dont think we can lie to Ignore. We could get erroneous results if the match glob is only supposed to match directories, and we are checking against a path pointing to a file.
> 
> https://github.com/BurntSushi/ripgrep/blob/67c0f576b6e1197c3cf80fe5d53cfd8326cf1d56/crates/ignore/src/gitignore.rs#L258-L266

That's why I was suggesting lying in the opposite direction. There is no "must be a file" gitignore glob type. So the worst case for lying here is a false _negative_ thing being ignored. Which would be ok here.

Reply:
There will be more changes... can follow up on it.

Reply:
Ah ok that makes sense! Also this is nice for us because we are using the parent path matcher, which ensures that things are directories https://github.com/BurntSushi/ripgrep/blob/67c0f576b6e1197c3cf80fe5d53cfd8326cf1d56/crates/ignore/src/gitignore.rs#L234 So the lie only even relevant for the initial path, which is much safer than if say someone uses a directory only pattern for pants.d and then all the events under pants.d *weren't* ignored.